**Домашнее задание к лекции «Случайные события. Случайные величины»**

In [ ]:
import numpy as np

Смоделировать игру против лотерейного автомата типа “777”. Игрок платит 1 руб., после чего выпадает случайное целое число, равномерно распределенное от 0 до 999. При некоторых значениях числа игрок получает выигрыш.

777: 200 руб.

999: 100 руб.

555: 50 руб.

333: 15 руб.

111: 10 руб.

*77: 5 руб.

**7: 3 руб.

*00: 2 руб.

**0: 1 руб.

*любая цифра

Вопросы:

Выгодна ли игра игроку?

Сколько в среднем приобретает или теряет игрок за одну игру?

**Дополнительное задание повышенной сложности. **

Теоретически рассчитать средний выигрыш (проигрыш) и сравнить с результатами моделирования



In [ ]:
from sre_constants import SUCCESS
# i - количество циклов, при стоимости одного цикла в 1р, i = потраченной сумме
i = 1000000
prize = 0
successes = 0
win_lst = {777: 200, 999:100, 555:50, 333:15, 111:10}
for _ in range(i):
  rez = np.random.randint(0, 999)
  if rez in win_lst:
    prize += win_lst[rez]
    successes += 1
  elif rez % 100 == 77:
    prize += 5
    successes += 1
  elif rez % 10 == 7:
    prize += 3
    successes += 1
  elif rez > 100:
    if rez % 100 == 0:
      prize += 2
      successes += 1
    elif rez % 10 == 0:
      prize += 1
      successes += 1
  elif rez < 100:
    if rez % 10 == 0:
      prize += 1
      successes += 1
ver = prize / i
p_practice = successes / i
print(f'Потрачено {i}руб. / игр проведено {i}')
print(f'Выиграно {prize}руб.')
print(f'Средний выигрыш за одну игру {ver:0.2f}руб.')
print(f'Игр выиграно {successes}')
print(f'Вероятность выигрыша {p_practice:0.2f}')

Потрачено 1000000руб. / игр проведено 1000000
Выиграно 696520руб.
Средний выигрыш за одну игру 0.70руб.
Игр выиграно 201379
Вероятность выигрыша 0.20


In [ ]:
# теоритический средний выигрыш
# так как у нас равновероятно выпадение любой комбинации, значит вероятность выпадения каждой комбинаций
# фиксированных комбинаций 111/333/555/777/999
p = 1/999
# комбинайции *77
p77 = 9 / 999   # 077/177/277/377/477/577/677/877/9777
# комбинайции **7
p7 = 90 / 999   # 10*10 - 9 - 1
# комбинайции **0
p0 = 90 / 999   # 9 * 10
# комбинайции *00
p00 = 9 / 999   # 100/200/300/400/500/600/700/800/900
aver_sum = (200 + 100 + 50 + 15 + 10 + 5 + 3 + 2 + 1) /9
ver_full = p*5 + p77 + p7 + p00 + p0    # теоритическая вероятность выигрыша
ver_sum = (p*200 + p*100 + p*50 + p*15 + p*10 + p77*5 + p7*3 + p00*2 + p0*1)
print(f'Теоритическая вероятность выигрыша {ver_full:0.2f}')
print(f'Теоритический выигрыш {ver_sum:0.2f}руб.')

Теоритическая вероятность выигрыша 0.20
Теоритический выигрыш 0.80руб.


Таки образом, мы получили:

*Теоритическая вероятность выигрыша - 0.20*

*Практическая вероятность выигрыша - 0.20*

*Теоритический выигрыш - 0.80руб.*

*Средний выигрыш за одну игру - 0.70руб.*

Можно сделать вывод о том, что игра на данном автомате заведомо проигрышна.
Теоритически, за каждую игру игрок тратит 1 руб, выигрывает - 0,8, соответственно теряет:

*(1 - 0,8) = 0,2руб*

На практике проигрыш еще больше:

*(1 - 0,7) = 0,3 руб*

